### ISSUE TITLE
Review implementation of MetaInfo[...]

### ISSUE THREAD

**Issue by hendrikweisser** [OpenMS/OpenMS#1759]
I'm currently struggling to find the cause of a segmentation fault in our code that handles meta data. This is not helped by the fact that parts of the implementation of the MetaInfo[...] classes are convoluted and don't seem sensible (to me).

As an example, `MetaInfo::exists` (a const method that should just check if a key exists) calls `MetaInfoRegistry::getIndex`, which calls `MetaInfoRegistry::registerName`. That function actually registers the key (changing the data structure!), but without setting a value, which apparently signifies that the key doesn't exist.

There's a lot of duplicated code between `MetaInfoRegistry::getIndex` and `MetaInfoRegistry::registerName`; I suspect that the implementation of `MetaInfoRegistry::getIndex` could be completely replaced by `return registerName(name, String::EMPTY, String::EMPTY);`.

In conclusion, the code should be reviewed at some point.

**Comment by hendrikweisser** [comment]
`MetaInfo::exists` contains a try/catch for `Exception::InvalidValue`, but I don't see where that exception could come from.

**Comment by hendrikweisser** [comment]
On a more fundamental level, MetaInfoRegistry uses several maps to store meta data keys and associated information (description and unit). Keys (strings) are first mapped to indexes (integers), which are then mapped to descriptions and units. In MetaInfo, the indexes are mapped to values.

Why the indirection via the indexes? At least as far as I'm aware, look-up of meta data is always based on the string keys. Is look-up of an integer in a map significantly faster than look-up of a string? If look-up speed is the primary concern, wouldn't a hash table work better?
If all else stays the same, wouldn't it be more economical to have on map in MetaInfoRegistry, mapping keys to structs of index/unit/description?

**Comment by timosachsenberg** [comment]
Hi. Looking up strings is much slower than indices  as the comparison (ordered) or hashing (unordered) take quite some time for strings. My guess is that someone introduced the indices because of performance issues. That way one can store the index externally and query it repeatedly (instead of the string).

**Comment by timosachsenberg** [comment]
Hash tables might have the unwanted property that order of keys are different if different implementations are used.

**Comment by timosachsenberg** [comment]
The idea of having a central registry is highly problematic in my view as it would require extensive locking to work in a muli-threading context. We don't do it which is a bad thing and a potential source of errors.

**Comment by timosachsenberg** [comment]
Hmm probably hard to get rid of the string to index mapping without storing a lot of strings in the individual objects that inherit MetaInfoInterface.

One improvement I can think about is to turn:
````    
    mutable std::map<UInt, String> index_to_name_;
    /// map from index to description
    mutable std::map<UInt, String> index_to_description_;
    /// map from index to unit
    mutable std::map<UInt, String> index_to_unit_;
```
into:
mutable std::vector<KeyDetail> index_to_key_detail;
where KeyDetail is a struct that contains name, description and unit.
KeyDetail.isEmpty() could delegate to return name.empty().
One big advantage is that the lock-up is constant time and the time spent blocking is greatly reduced. 

also the map from name to index could be a boost::unordered_map.

Both vector and unordered_map could be initialized to a reasonable size to reduce reallocation during locking.

**Comment by hendrikweisser** [comment]
> Looking up strings is much slower than indices as the comparison (ordered) or hashing (unordered) take quite some time for strings.

Fair point.

> Hash tables might have the unwanted property that order of keys are different if different implementations are used.

Why would that be a problem?

> `mutable std::vector<KeyDetail> index_to_key_detail;`

I was thinking `std::map<String, KeyDetail> registered_names` (or `boost::unordered_map`) in MetaInfoRegistry, with KeyDetail storing index/unit/description (as I said above), and perhaps `vector<DataValue> index_to_value` in MetaInfo. There's a runtime/memory trade-off between using a vector (essentially a hash table where keys are ints and the hash function is identity) and using a map in MetaInfo.

My rationale is that querying for meta data (almost?) always uses the string keys, rather than the indexes. So in practice you have to do one string look-up per query anyway, and I think our aim shouldn't be to avoid that one look-up (i.e. provide an index-based interface), but to avoid doing the look-up more times than necessary.

**Comment by hendrikweisser** [comment]
One other thing that I think we should definitely do is get rid of the "mutable" declarations in MetaInfoRegistry. Unless I'm mistaken, they're only there because `getIndex` (const) calls `registerName` (not const, but declared as such), which I'm pretty sure could be avoided with a slightly improved design. In particular, `getIndex(name)` should *not* register `name` as a key, but return "not found" in some form or another if the key doesn't exist.

**Comment by timosachsenberg** [comment]
The vector should be faster and less memory consuming as sorting only very rarely happens.

**Comment by hendrikweisser** [comment]
> The vector should be faster and less memory consuming as sorting only very rarely happens.

Which one do you mean?

**Comment by timosachsenberg** [comment]
are you on gitter? its easier to chat there https://gitter.im/OpenMS/OpenMS?utm_source=badge&utm_medium=badge&utm_campaign=pr-badge

**Comment by timosachsenberg** [comment]
In MetaInfo we can't replace:
```
std::map<UInt, DataValue> index_to_value_; by vector<DataValue> index_to_value
```
as we need to keep memory consumption as low as possible

**Comment by hendrikweisser** [comment]
I was just coming to that. I think we should reconsider this bit:

> When registering a new name an index >= 1024 is assigned. Indices from 1 to 1023 are reserved for fast access and will never change: ...

That must have seemed like a good idea in the early days of OpenMS, but I think experience has shown that the meta data system isn't really used in this way (i.e. tools registering their keys and then using index-based access). Especially if we want to turn the index-to-value map in MetaInfo into a vector, we might not want to reserve all that space unnecessarily. Conversely, if we don't reserve any unused keys, we don't consume more memory by using a vector instead of a map.

**Comment by hendrikweisser** [comment]
What is the reason for all the `#pragma omp critical (MetaInfoRegistry)` lines in MetaInfoRegistry?

**Comment by timosachsenberg** [comment]
it aquires a lock (name MetaInfoRegistry) used by methods that read and write the datastructure

**Comment by hendrikweisser** [comment]
Is the lock actually required for reading, or only for modifying?

**Comment by timosachsenberg** [comment]
for both. writing in another thread might invalidate iterators needed during traversing the map

**Comment by hendrikweisser** [comment]
Fix for the "mutable"/const'ness problems in #1778.

**Comment by hendrikweisser** [comment]
Additional improvement (reduced memory overhead) made by Timo in #3733.
